File: TestApp/app/src/main/java/com/tsi84/testapp/SecondActivity.java
package com.tsi84.testapp;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Handler;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;

import com.tsi84.testapp.model.ConstVariables;

public class SecondActivity extends BaseActivity implements View.OnClickListener {
    private static final String TAG = SecondActivity.class.getSimpleName();

    private ImageView mImageSample;
    private Button mBtnStart;
    private Button mBtnClose;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_second);
        mBtnStart = findViewById(R.id.btn_start);
        mBtnStart.setOnClickListener(this);
        mBtnClose = findViewById(R.id.btn_close);
        mBtnClose.setOnClickListener(this);
        mImageSample = findViewById(R.id.img_sample);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btn_start:
                updateImage();
//                updateImageRunnable04();
//                UpdateImageTask task = new UpdateImageTask();
//                task.execute(R.drawable.flo);
                break;

            case R.id.btn_close:
                finish();
                break;
        }
    }

    private void updateImage() {
        final Bitmap bitmap = BitmapFactory.decodeResource(getResources(), R.drawable.flo);
        mImageSample.setImageBitmap(bitmap);
    }

    private void updateImageRunnable01() {
        new Thread(new Runnable() {
            public void run() {
                // a potentially time consuming task
                final Bitmap bitmap = BitmapFactory.decodeResource(getResources(), R.drawable.flo);
                mImageSample.post(new Runnable() {
                    public void run() {
                        mImageSample.setImageBitmap(bitmap);
                    }
                });
            }
        }).start();
    }

    private void updateImageRunnable02() {
        Handler handler = new Handler();
        handler.post(new Runnable() {
            @Override
            public void run() {
                final Bitmap bitmap = BitmapFactory.decodeResource(getResources(), R.drawable.flo);
                mImageSample.setImageBitmap(bitmap);
            }
        });
    }

    private void updateImageRunnable03() {
        this.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                final Bitmap bitmap = BitmapFactory.decodeResource(getResources(), R.drawable.flo);
                mImageSample.setImageBitmap(bitmap);
            }
        });
    }

    private void updateImageRunnable04() {
        showLoadingDialog();
        Handler handler = new Handler();
        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                final Bitmap bitmap = BitmapFactory.decodeResource(getResources(), R.drawable.flo);
                mImageSample.setImageBitmap(bitmap);
                hideLoadingDialog();
            }
        }, ConstVariables.VALUE_OF_MAX_LOADING);
    }

    private class UpdateImageTask extends AsyncTask<Integer, String, Bitmap> {

        public UpdateImageTask() {

        }

        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            showLoadingDialog();
        }

        @Override
        protected Bitmap doInBackground(Integer... params) {
            int resourceId = params[0];

            // progress UI
            publishProgress(getResources().getString(R.string.message_image_loading));

            // background process
            final Bitmap bitmap = BitmapFactory.decodeResource(getResources(), resourceId);

            // test sleep..
            try {
                Thread.sleep(3000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            return bitmap;
        }

        @Override
        protected void onProgressUpdate(String... values) {
            super.onProgressUpdate(values);

            setLoadingDialogText(values[0]);
        }

        @Override
        protected void onPostExecute(Bitmap bitmap) {
            super.onPostExecute(bitmap);
            hideLoadingDialog();
            if (bitmap != null) {
                mImageSample.setImageBitmap(bitmap);
            }
        }
    }
}

File: TestKotlin/app/src/main/java/com/tsi84/testkotlin/SecondActivity.kt
package com.tsi84.testkotlin

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.AsyncTask
import android.os.Bundle
import android.os.Handler
import android.view.View
import android.widget.Button
import android.widget.ImageView
import com.tsi84.testkotlin.model.ConstVariables

class SecondActivity : BaseActivity(), View.OnClickListener {
    private var mImageSample: ImageView? = null
    private var mBtnStart: Button? = null
    private var mBtnClose: Button? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_second)
        mBtnStart = findViewById(R.id.btn_start)
        mBtnStart?.setOnClickListener(this)
        mBtnClose = findViewById(R.id.btn_close)
        mBtnClose?.setOnClickListener(this)
        mImageSample = findViewById(R.id.img_sample)
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.btn_start -> {
                // normal
//                updateImage();
                // runnable
//                updateImageRunnable04();

                // AsyncTask
                val task = UpdateImageTask()
                task.execute(R.drawable.flo)
            }
            R.id.btn_close -> finish()
        }
    }

    private fun updateImage() {
        val bitmap = BitmapFactory.decodeResource(resources, R.drawable.flo)
        mImageSample!!.setImageBitmap(bitmap)
    }

    private fun updateImageRunnable01() {
        Thread(Runnable { // a potentially time consuming task
            val bitmap = BitmapFactory.decodeResource(resources, R.drawable.flo)
            mImageSample!!.post { mImageSample!!.setImageBitmap(bitmap) }
        }).start()
    }

    private fun updateImageRunnable02() {
        val handler = Handler()
        handler.post {
            val bitmap = BitmapFactory.decodeResource(resources, R.drawable.flo)
            mImageSample!!.setImageBitmap(bitmap)
        }
    }

    private fun updateImageRunnable03() {
        runOnUiThread {
            val bitmap = BitmapFactory.decodeResource(resources, R.drawable.flo)
            mImageSample!!.setImageBitmap(bitmap)
        }
    }

    private fun updateImageRunnable04() {
        showLoadingDialog()
        val handler = Handler()
        handler.postDelayed({
            val bitmap = BitmapFactory.decodeResource(resources, R.drawable.flo)
            mImageSample!!.setImageBitmap(bitmap)
            hideLoadingDialog()
        }, ConstVariables.VALUE_OF_MAX_LOADING.toLong())
    }

    private inner class UpdateImageTask : AsyncTask<Int?, String?, Bitmap?>() {
        override fun onPreExecute() {
            super.onPreExecute()
            showLoadingDialog()
        }

        override fun doInBackground(vararg p0: Int?): Bitmap? {
            val resourceId = p0.get(0)

            // progress UI
            publishProgress(resources.getString(R.string.message_image_loading))

            // background process
            val bitmap = resourceId?.let { BitmapFactory.decodeResource(resources, it) }

            // test sleep..
            try {
                Thread.sleep(3000)
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
            return bitmap
        }

        override fun onProgressUpdate(vararg p0: String?) {
            super.onProgressUpdate(*p0)
            setLoadingDialogText(p0[0])
        }

        override fun onPostExecute(bitmap: Bitmap?) {
            super.onPostExecute(bitmap)
            hideLoadingDialog()
            if (bitmap != null) {
                mImageSample!!.setImageBitmap(bitmap)
            }
        }
    }

    companion object {
        private val TAG = SecondActivity::class.java.simpleName
    }
}